src/ott/solvers/linear/univariate.py

```diff
@@ -11,336 +11,311 @@
-    entries, but sums to 1.0
+    The univariate output. Note that the mass can be null in some entries,
+    but sums to :math:`1`.
 
   Notes:
-    Inspired by :func:`~scipy.stats.wasserstein_distance`,
+    This function was inspired by :func:`~scipy.stats.wasserstein_distance`,
     but can be used with other costs, not just :math:`c(x, y) = |x - y|`.
-  """
-  x, i_x = mu.sort_and_argsort(x, argsort=True)
-  y, i_y = mu.sort_and_argsort(y, argsort=True)
-
-  all_values = jnp.concatenate([x, y])
-  all_values_sorted, all_values_sorter = mu.sort_and_argsort(
-      all_values, argsort=True
+  """  # noqa: E501
+
+  @functools.partial(jax.vmap, in_axes=[1, 1])
+  def dist(x: jnp.ndarray, y: jnp.ndarray):
+    x, i_x = mu.sort_and_argsort(x, argsort=True)
+    y, i_y = mu.sort_and_argsort(y, argsort=True)
+
+    all_values = jnp.concatenate([x, y])
+    all_values_sorted, all_values_sorter = mu.sort_and_argsort(
+        all_values, argsort=True
+    )
+
+    x_pdf = jnp.concatenate([prob.a[i_x], jnp.zeros_like(prob.b)])
+    x_pdf = x_pdf[all_values_sorter]
+    y_pdf = jnp.concatenate([jnp.zeros_like(prob.a), prob.b[i_y]])
+    y_pdf = y_pdf[all_values_sorter]
+
+    x_cdf = jnp.cumsum(x_pdf)
+    y_cdf = jnp.cumsum(y_pdf)
+
+    x_y_cdfs = jnp.concatenate([x_cdf, y_cdf])
+    quantile_levels, _ = mu.sort_and_argsort(x_y_cdfs, argsort=False)
+
+    i_x_cdf_inv = jnp.searchsorted(x_cdf, quantile_levels)
+    x_cdf_inv = all_values_sorted[i_x_cdf_inv]
+    i_y_cdf_inv = jnp.searchsorted(y_cdf, quantile_levels)
+    y_cdf_inv = all_values_sorted[i_y_cdf_inv]
+
+    diff_q = jnp.diff(quantile_levels)
+    successive_costs = jax.vmap(prob.geom.cost_fn.h)(
+        x_cdf_inv[1:, None] - y_cdf_inv[1:, None]
+    )
+    cost = jnp.sum(successive_costs * diff_q)
+
+    if not return_transport:
+      return cost, None, None
+
+    n = x.shape[0]
+    i_in_sorted_x_of_quantile = all_values_sorter[i_x_cdf_inv] % n
+    i_in_sorted_y_of_quantile = all_values_sorter[i_y_cdf_inv] - n
+    orig_i = i_x[i_in_sorted_x_of_quantile][1:]
+    orig_j = i_y[i_in_sorted_y_of_quantile][1:]
+    paired_indices, mass_paired_indices = jnp.stack([orig_i, orig_j]), diff_q
+
+    return cost, paired_indices, mass_paired_indices
+
+  ot_costs, pi, mpi = dist(prob.geom.x, prob.geom.y)
+  return UnivariateOutput(
+      prob,
+      ot_costs,
+      paired_indices=pi,
+      mass_paired_indices=mpi,
   )
 
-  x_pdf = jnp.concatenate([a[i_x], jnp.zeros_like(b)])[all_values_sorter]
-  y_pdf = jnp.concatenate([jnp.zeros_like(a), b[i_y]])[all_values_sorter]
 
-  x_cdf = jnp.cumsum(x_pdf)
-  y_cdf = jnp.cumsum(y_pdf)
+def north_west_distance(prob: linear_problem.LinearProblem) -> UnivariateOutput:
+  r"""Distance between two distributions using the north-west corner rule.
 
-  x_y_cdfs = jnp.concatenate([x_cdf, y_cdf])
-  quantile_levels, _ = mu.sort_and_argsort(x_y_cdfs, argsort=False)
+  This function uses the dual form as described in :cite:`sejourne:22`, alg. 3.
```

Comment:
I don't think `dual form` means much here. Maybe reference NW solution in 3.4.2 computational OT book + Séjourné to also store dual variables?